### PR TITLE
Use syntax kinds in more cases.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
ISyntaxKindsService exists so we can easily abstract features over C#/VB syntax cases that are effectively equivalent at the AST level.  This PR fixes up several features that should have been using this service but were instead rolling their own mechanism to operate at an abstract level over syntax.

I also NRT enabled all the files i touched to help move things forward.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide @mavasani .  General cleanup and simplification of features by moving AST logic down to our common core.

File: src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -14,8 +17,10 @@ internal class CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer
             SyntaxKind,
             AnonymousObjectCreationExpressionSyntax>
     {
-        protected override SyntaxKind GetAnonymousObjectCreationExpressionSyntaxKind()
-            => SyntaxKind.AnonymousObjectCreationExpression;
+        public CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer()
+            : base(CSharpSyntaxKindsService.Instance)
```
Comment (mavasani) [inline src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs]: So the assumption here is eventually we will move ISyntaxFactsService and ISyntaxKindsService down to the shared layer where these analyzers moved to CodeStyle layer will still be able to use them? Sounds reasonable to me, but just confirming.

File: src/Features/CSharp/Portable/UseNamedArguments/CSharpUseNamedArgumentsCodeRefactoringProvider.cs [hunk 2]
```diff
@@ -21,15 +23,12 @@ private abstract class BaseAnalyzer<TSyntax, TSyntaxList> : Analyzer<TSyntax, TS
             where TSyntax : SyntaxNode
             where TSyntaxList : SyntaxNode
         {
-            protected sealed override SyntaxNode GetReceiver(SyntaxNode argument)
-                => argument.Parent.Parent;
+            protected sealed override SyntaxNode? GetReceiver(SyntaxNode argument)
+                => argument.Parent?.Parent;
```
Comment (mavasani) [inline src/Features/CSharp/Portable/UseNamedArguments/CSharpUseNamedArgumentsCodeRefactoringProvider.cs]: Nice!

File: src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs]: Yup.  these are pure-data-singletons.  They can move down very effectively.  There may be some places where they're not pure-data (i.e. they use other imported-services).  but those shoudl be very rare, and can be broken out.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ConvertAnonymousTypeToTuple/CSharpConvertAnonymousTypeToTupleDiagnosticAnalyzer.cs]: 👍

File: src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -110,10 +115,9 @@ private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, INamedTypeSymbol e
                     syntaxFacts.GetOperandOfPrefixUnaryExpression(conditionNode));
             }
 
-            var isEqualityLikeCondition = TryAnalyzeCondition(
-                context, syntaxFacts, referenceEqualsMethodOpt, conditionNode,
-                out var conditionPartToCheck, out var isEquals);
-            if (!isEqualityLikeCondition)
+            if (!TryAnalyzeCondition(
+                    context, syntaxFacts, referenceEqualsMethodOpt, conditionNode,
+                    out var conditionPartToCheck, out var isEquals))
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs]: needed for NRT to work.  if not directly in the 'if' check, the flow analysis doesn't know what is assigned or not later on.

File: src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -188,8 +192,12 @@ private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, INamedTypeSymbol e
...
 
         private bool TryAnalyzeCondition(
-            SyntaxNodeAnalysisContext context, ISyntaxFactsService syntaxFacts, IMethodSymbol referenceEqualsMethodOpt, SyntaxNode conditionNode,
-            out SyntaxNode conditionPartToCheck, out bool isEquals)
+            SyntaxNodeAnalysisContext context,
+            ISyntaxFactsService syntaxFacts,
+            IMethodSymbol? referenceEqualsMethodOpt,
+            SyntaxNode conditionNode,
+            [NotNullWhen(true)] out SyntaxNode? conditionPartToCheck,
+            out bool isEquals)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs]: wrapped just to add NotNullWhen attribute

File: src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs [hunk 5]
```diff
@@ -227,8 +237,12 @@ private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, INamedTypeSymbol e
...
 
         private static bool TryAnalyzeInvocationCondition(
-            SyntaxNodeAnalysisContext context, ISyntaxFactsService syntaxFacts, IMethodSymbol referenceEqualsMethodOpt, TInvocationExpression invocation,
-            out SyntaxNode conditionPartToCheck, out bool isEquals)
+            SyntaxNodeAnalysisContext context,
+            ISyntaxFactsService syntaxFacts,
+            IMethodSymbol? referenceEqualsMethodOpt,
+            TInvocationExpression invocation,
+            [NotNullWhen(true)] out SyntaxNode? conditionPartToCheck,
+            out bool isEquals)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs]: wrapped just to add NotNullWhen attribute

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs [hunk 6]
```diff
@@ -443,13 +412,13 @@ public bool IsIdentifierEscapeCharacter(char c)
         public bool IsValidIdentifier(string identifier)
         {
             var token = SyntaxFactory.ParseToken(identifier);
-            return IsIdentifier(token) && !token.ContainsDiagnostics && token.ToString().Length == identifier.Length;
+            return this.IsIdentifier(token) && !token.ContainsDiagnostics && token.ToString().Length == identifier.Length;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs]: `IsIdentifier` became an extension, and you are required to call an extension off of a non-implicit receiver in C#.

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/AbstractSyntaxFactsService.cs [hunk 7]
```diff
@@ -158,8 +158,12 @@ protected AbstractSyntaxFactsService()
                     Matcher.Repeat(singleBlankLine));
         }
 
-        public abstract bool IsWhitespaceTrivia(SyntaxTrivia trivia);
-        public abstract bool IsEndOfLineTrivia(SyntaxTrivia trivia);
+        private bool IsWhitespaceTrivia(SyntaxTrivia trivia)
+            => SyntaxKinds.WhitespaceTrivia == trivia.RawKind;
+
+        private bool IsEndOfLineTrivia(SyntaxTrivia trivia)
+            => SyntaxKinds.EndOfLineTrivia == trivia.RawKind;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/AbstractSyntaxFactsService.cs]: couldn't move due to some weird inheritance shenanigans.  Will fixup as more cleanup happens.  but there is already a ton of goodness in this PR.

**Comment by CyrusNajmabadi** [comment]
@mavasani Any issues with this going in?

**Comment by mavasani** [comment]
@CyrusNajmabadi I haven't reviewed the whole PR yet. Let me get to this later today - I don't really see any issues here.

**Comment by CyrusNajmabadi** [comment]
Ok @mavasani .  I've merged in your changes.  No pain with them so far in this PR.  So it should be easy to get in.  Thanks! :)

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs [hunk 8]
```diff
@@ -29,9 +29,6 @@ internal interface ISyntaxFactsService : ILanguageService
 
         SyntaxToken ParseToken(string text);
 
-        bool IsAwaitKeyword(SyntaxToken token);
```
Comment (mavasani) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs]: I am presuming these all just move to SyntaxKinds service?

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/SyntaxFactsService/ISyntaxFactsServiceExtensions.cs [hunk 9]
```diff
@@ -183,5 +186,141 @@ public static void GetPartsOfMemberAccessExpression(this ISyntaxFactsService syn
 
         public static void GetPartsOfConditionalAccessExpression(this ISyntaxFactsService syntaxFacts, SyntaxNode node, out SyntaxNode expression, out SyntaxNode whenNotNull)
             => syntaxFacts.GetPartsOfConditionalAccessExpression(node, out expression, out _, out whenNotNull);
+
+        #region ISyntaxKinds forwarding methods
+
+        public static bool IsEndOfLineTrivia(this ISyntaxFactsService syntaxFacts, SyntaxTrivia trivia)
```
Comment (mavasani) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/SyntaxFactsService/ISyntaxFactsServiceExtensions.cs]: Ah they are all moved here. Looks fine.

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs [hunk 10]
```diff
@@ -284,7 +284,7 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
         {
             localSymbol = semanticModel.GetSymbolInfo(identifier).Symbol as ILocalSymbol;
             declarator = localSymbol?.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax() as VariableDeclaratorSyntax;
-            return declarator != null;
+            return localSymbol != null && declarator != null;
```
Comment (mavasani) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: Why was this needed? Seems `declarator` cannot be non-null unless `localSymbol` is also non-nulll.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: hrmm..  compiler doesn't seem smart enough here and thinks its posisble that localSymbol might not be assigned on out  here.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs]: yup

**Comment by CyrusNajmabadi** [comment]
Test failure is Microsoft.CodeAnalysis.CSharp.UnitTests.Emit.DeterministicTests.CompareAllBytesEmitted_Debug 

Not sure what we do about that @mavasani @jasonmalinowski

**Comment by CyrusNajmabadi** [comment]
@mavasani I need a little help with this PR.  I thought i would try the exercise of making this simple service available in the CodeStyle layer just as pure data, and in the CodeFixes and Workspace layers as an actual ILanguageService.

I've tried for about an hour and I can't seem to make any headway here.  I can't seem to figure out how to use the shared files to make this possible.

First, some things i just don't understand:

image

What are these projects?  They don't seem to actually contain anything or do anything.  I really don't know what purpose they serve.

Second, CodeStyle references these projects:

image

As mentioned above, i don't actually know what `Analyzers` is.  I couldn't seem to do anything with it.  So i attempted to move `ISyntaxKindsService` into `CompilerExtensions`.  I needed to split it into two parts.  One that just had the data, and one that extended `ILanguageService`.  I then attempted to keep the second part just in `WorkspaceExtensions`.  

The problem here is that the CodeStyle.Fixes project references the CodeStyle project *and* the `WorkspaceExtensions` project.   So there's now an ambiguity because types get pulled in through two paths.  

I'm really not sure what the right way is to approach this.   Could you advise here?

**Comment by mavasani** [comment]
Analyzers, CodeFixes and corresponding unit test projects are empty shared projects that were split out from https://github.com/dotnet/roslyn/pull/41363. That PR moves a few IDE analyzers, code fixers and their unit tests into each of these projects. I would suggest checking out my branch from that PR to locally see the structure.

> So i attempted to move ISyntaxKindsService into CompilerExtensions. I needed to split it into two parts. One that just had the data, and one that extended ILanguageService. I then attempted to keep the second part just in WorkspaceExtensions.

I think that is the perfect approach, and I would have attempted the same.

> The problem here is that the CodeStyle.Fixes project references the CodeStyle project and the WorkspaceExtensions project. So there's now an ambiguity because types get pulled in through two paths.

I hit exactly the same problem, see my comment here: https://github.com/dotnet/roslyn/pull/41363/files#r376442419. In that PR I have moved all the public Option related types that are explicitly linked into CodeStyle project for analyzers (such as OptionSet, IOption, `Option<T>`, CodeStyleOption, CodeStyleOptions, NotificationOptions, etc.) into a separate namespace with an `.Internal` sub-namespace so that files in WorkspaceExtensions.shproj can appropriately reference the types from CodeStyle project or from Workspaces project when imported in the respective projects by having using statements conditioned on `CODE_STYLE` preprocessor directive. This is unfortunate side-effect of the fact that our IDE analyzers need to reference these types, but cannot reference Workspaces. I don't see a good way around this except the approach I have taken in that PR.
Note that Sam has been using extern aliases to handle similar conflicts in shared formatter code, but we cannot do so for the entire Code Style layer because VB does not support them, and personally I find them quite difficult to fathom. I'd rather have the duplicated types in CodeStyle layer be duplicated into a separate namespace.
My PR also adds a `BannedSymbols.txt` for CodeStyle layer: https://github.com/dotnet/roslyn/pull/41363/files#diff-8747d0acadea5b353b1871784c14003c. This ensures that the shared files in WorkspaceExtensions.shproj do not accidentally reference the Option types from Workspaces layer - you should see an analyzer diagnostic on wrong API usage once that PR goes in.

**Comment by mavasani** [comment]
Also see https://github.com/dotnet/roslyn/pull/41363/files#r376523357

**Comment by mavasani** [comment]
Seems like the last link does not work directly. I meant to point you to the changes to `CodeStyleOption.cs` in that PR and the comment I added on that file:

> Namespace changes for all options related types are needed to avoid use of extern aliases in CodeFixes and UniTests project which have access to both the core CodeStyle project and Workspaces assembly/project where they are defined.

**Comment by CyrusNajmabadi** [comment]
> I think that is the perfect approach, and I would have attempted the same.

I've gone with another approach that works well (whcih you can see in this PR).

First, i'm moving *only* the data portions of these APIs into the shared-projects.  That's all that is contained there.  It makes the data-apis available to `CodeStyle`, `CodeStyle.Fixes` and `Workspaces (and above)`.  Then, i make the `ILanguageService` portion *only* available in the `Workspaces (and above)` layer.  The idea here is that `CodeStyle` and `CodeStyle.Fixes` are never trying to work with these language services.  They are only ever working with simple data apis/utilities/extensions.

It keeps a whole layer of complexity out.

The only downside is that it then means that the CodeFixes *must* have a C# and VB side if they need to reference these data APIs.  In general, this simply means an abstract base class, with derived classes that pass up the data-apis to the base class.  The base class can't use `document.GetLanguageService` but instead get the services from the derived types.

I think this is likely just fine for our needs here.

Thoughts @mavasani ?

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/VisualBasic/VisualBasicWorkspaceExtensions.projitems [hunk 11]
```diff
@@ -60,8 +60,6 @@
     <Compile Include="$(MSBuildThisFileDirectory)LanguageServices\VisualBasicSymbolDeclarationService.vb" />
     <Compile Include="$(MSBuildThisFileDirectory)LanguageServices\VisualBasicSyntaxFactsService.vb" />
     <Compile Include="$(MSBuildThisFileDirectory)LanguageServices\VisualBasicSyntaxFactsServiceFactory.vb" />
-    <Compile Include="$(MSBuildThisFileDirectory)LanguageServices\VisualBasicSyntaxKindsService.vb" />
-    <Compile Include="$(MSBuildThisFileDirectory)LanguageServices\VisualBasicSyntaxKindsServiceFactory.vb" />
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/VisualBasic/VisualBasicWorkspaceExtensions.projitems]: VisualBasicSyntaxKindsService moved into VisualBasicCompilerExtensions (as a pure data api).
VisualBasicSyntaxKindsServiceFactory moved into VisualBasicWorkspace (as a workspace/mef/service discovery facet).

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFacts/ISyntaxKindsService_LanguageService.cs [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis.LanguageServices
+{
+    /// <summary>
+    /// Provides a uniform view of SyntaxKinds over C# and VB for constructs they have
+    /// in common.
+    /// </summary>
+    internal partial interface ISyntaxKindsService : ILanguageService
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/LanguageServices/SyntaxFacts/ISyntaxKindsService_LanguageService.cs]: I think we should still consider having this in the shared WorkspaceExtensions layer. CodeFixes themselves can easily get the singleton implementation of this, but it gets pretty messy when we try to move down some extension methods/utilities written in language-agnostic Workspaces layer into the language-agnostic WorkspaceExtensions layer. We need to change all call sites and method chain to thread through the instance of `ISyntaxKindsService` to that helper.

My previous PR (merged yesterday) has already enabled support for `ILanguageService` detection using `document.GetLanguageService` in WorkspaceExtensions.shproj and CodeStyleFixes projects and above. My remaining PR adds MEF based detection of `ILanguageService` using a special CodeStyle layer implementation of HostLanguageServices. I think once that PR goes in, it should be fine to move all the common language services, such as this one, into WorkspaceExtensions.shproj as an `ILanguageService` itself. Does that seem reasonable?

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/VisualBasic/VisualBasicWorkspaceExtensions.projitems (hunk 11 above)
Comment (mavasani) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/VisualBasic/VisualBasicWorkspaceExtensions.projitems]: Nice!

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/VisualBasicCompilerExtensions.projitems [hunk 13]
```diff
@@ -13,5 +13,6 @@
     <Compile Include="$(MSBuildThisFileDirectory)Extensions\SyntaxTreeExtensions_SharedWithCodeStyle.vb" />
     <Compile Include="$(MSBuildThisFileDirectory)Extensions\SyntaxTriviaExtensions.vb" />
     <Compile Include="$(MSBuildThisFileDirectory)Helpers\RemoveUnnecessaryImports\VisualBasicUnnecessaryImportsProvider.vb" />
+    <Compile Include="$(MSBuildThisFileDirectory)Services\SyntaxFacts\VisualBasicSyntaxKindsService.vb" />
```
Comment (mavasani) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/VisualBasicCompilerExtensions.projitems]: Maybe name the folder `LanguageServices` for consistency with other projects.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/CSharpCompilerExtensions.projitems [hunk 14]
```diff
@@ -29,6 +29,7 @@
     <Compile Include="$(MSBuildThisFileDirectory)Extensions\SyntaxTriviaListExtensions.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Extensions\TypeSyntaxExtensions.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Helpers\RemoveUnnecessaryImports\CSharpUnnecessaryImportsProvider.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)Services\SyntaxFacts\CSharpSyntaxKindsService.cs" />
```
Comment (mavasani) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/CSharpCompilerExtensions.projitems]: Maybe name the folder `LanguageServices` instead of `Services` for consistency with other projects.

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFacts/ISyntaxKindsService_LanguageService.cs (hunk 12 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/LanguageServices/SyntaxFacts/ISyntaxKindsService_LanguageService.cs]: > Does that seem reasonable?

So my preference would be to get these things moved down.  Then, when we're ready to have a simple approachthat allows hte services to be used effectively at CodeStyle/CodeStyle.Fixes/Worskspaces layer, we just make that change as appropriately.

I feel like we keep on trying to push too much stuff into really large changes.  I would far prefer smaller, indepedent changes that focus on solving one case.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/VisualBasicCompilerExtensions.projitems (hunk 13 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/VisualBasicCompilerExtensions.projitems]: So this was intentional.  At this level it is *not* a LanguageService :)  The LanguageService stuff is introduced higher up.  If we can push language services down later to this level, then we should rename this folder.

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFacts/ISyntaxKindsService_LanguageService.cs (hunk 12 above)
Comment (mavasani) [reply src/Workspaces/Core/Portable/LanguageServices/SyntaxFacts/ISyntaxKindsService_LanguageService.cs]: Absolutely, your change as is seems perfectly fine. I was just stating something that can be done in future.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/LanguageServices/SyntaxFacts/ISyntaxKindsService_LanguageService.cs]: Note: i would also strongly prefer not having to have any sort of `#ifdefs` or other mechanisms to work around sharing issues.  (i'm fine with them for versioning issues, like the NRT api stuff).  to me, we should just be tryin to get it so that the files just link in the exact same way into all locations they're needed.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/VisualBasicCompilerExtensions.projitems (hunk 13 above)
Comment (mavasani) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/VisualBasicCompilerExtensions.projitems]: Ah, yes that is correct.

**Comment by CyrusNajmabadi** [comment]
@mavasani @jasonmalinowski @tmat tehre seems to be an emit test that continues to be flakey.  (i've seen like 50% of my PRs die on that test).  Can we do something about it?  Thanks! :)

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFacts/ISyntaxKindsService_LanguageService.cs (hunk 12 above)
Comment (mavasani) [reply src/Workspaces/Core/Portable/LanguageServices/SyntaxFacts/ISyntaxKindsService_LanguageService.cs]: Yes, this cleanup is tracked in the follow-up item. Note that those conditional `#ifdefs` will be unavoidable for any code in shared layer that references any Options related type - these are public types on Workspaces and must be linked into the core CodeStyle layer for our IDE code style analyzers to be moved there. This means shared code dealing with Options related types will need to choose between the appropriate definition from these assemblies based on `#if CODE_STYLE` statements. Apart from these files, we should not require any `#ifdefs`

**Comment by tmat** [comment]
@CyrusNajmabadi That's not good. This test is testing compiler determinism and the fact that's failing means there is a bug in the compiler. 

> Can we do something about it?

Yes. You can start by filing a bug on the compiler.

File: src/Workspaces/Core/Portable/LanguageServices/SyntaxFacts/ISyntaxKindsService_LanguageService.cs (hunk 12 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/LanguageServices/SyntaxFacts/ISyntaxKindsService_LanguageService.cs]: SGTM.  Thanks! :)

**Comment by CyrusNajmabadi** [comment]
Done!  https://github.com/dotnet/roslyn/issues/41617

**Comment by CyrusNajmabadi** [comment]
@mavasani This is ready to go in.  Failing emit test is tracked with  #41617

Thanks!

**Comment by CyrusNajmabadi** [comment]
Thanks!
